lib/rust/parser/src/main.rs

```diff
@@ -0,0 +1,835 @@
+                matched_macro_def,
+            };
+            mem::swap(&mut self.current_macro, &mut current_macro);
+            self.macro_stack.push(current_macro);
+            ResolverStep::NewSegmentStarted
+        } else {
+            event!(TRACE, "Consuming token as current segment body.");
+            ResolverStep::NormalToken
+        }
+    }
+
+    fn enter(
+        matched_macro_def: &mut Option<Rc<macros::Definition<'a>>>,
+        path: &[PartiallyMatchedMacro<'a>],
+    ) -> MacroMatchTree<'a> {
+        *matched_macro_def = None;
+        let mut new_section_tree = MacroMatchTree::default();
+        for v in path {
+            if let Some(first) = v.required_segments.head() {
+                let tail = v.required_segments.tail().cloned().unwrap_or_default();
+                let definition = v.definition.clone_ref();
+                let x = PartiallyMatchedMacro { required_segments: tail, definition };
+                if let Some(node) = new_section_tree.get_mut(&first.header) {
+                    node.push(x);
+                } else {
+                    new_section_tree.insert(first.header, NonEmptyVec::singleton(x));
+                }
+            } else {
+                if matched_macro_def.is_some() {
+                    event!(ERROR, "Internal error. Duplicate macro definition.");
+                }
+                *matched_macro_def = Some(v.definition.clone_ref());
+            }
+        }
+        new_section_tree
+    }
+}
+
+
+// FIXME: hardcoded values + not finished implementation.
+pub fn precedence_of(operator: &str) -> usize {
+    match operator {
+        "+" => 3,
+        "-" => 3,
+        "*" => 7,
+        _ => panic!("Operator not supported: {}", operator),
+    }
+}
+
+#[derive(Clone, Copy, Debug, Deref, DerefMut)]
+pub struct WithPrecedence<T> {
+    #[deref]
+    #[deref_mut]
+    elem:       T,
+    precedence: usize,
+}
+
+impl<T> WithPrecedence<T> {
+    pub fn new(precedence: usize, elem: T) -> Self {
+        Self { elem, precedence }
+    }
+}
+
+
+fn annotate_tokens_that_need_spacing(items: Vec<syntax::Item>) -> Vec<syntax::Item> {
+    items
+        .into_iter()
+        .map(|item| match item {
+            syntax::Item::Token(_) => item,
+            syntax::Item::Tree(ast) => match &ast.elem {
+                syntax::tree::Type::MultiSegmentApp(data) => {
+                    if data.segments.first().header.elem.variant() != token::TypeVariant::Symbol {
+                        syntax::Item::Tree(
+                            ast.with_error(
+                                "This expression cannot be used in a non-spaced equation.",
+                            ),
+                        )
+                    } else {
+                        syntax::Item::Tree(ast)
+                    }
+                }
+                _ => syntax::Item::Tree(ast),
+            },
+        })
+        .collect()
+}
```

Comment:
Definitely. There are more places like that in `main.rs`. That's why I've written in the PR description that `main.rs` contains a temporary code and that it will be improved and refactored to other modules in the next PRs.